SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,276 @@
+{
+	LocalVirtualWorkerId = InLocalVirtualWorkerId;
+	for (const auto& Elem : LayerNameToLBStrategy)
+	{
+		Elem.Value->SetLocalVirtualWorkerId(InLocalVirtualWorkerId);
+	}
+}
+
+bool ULayeredLBStrategy::ShouldHaveAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy not ready to relinquish authority for Actor %s."), *AActor::GetDebugName(&Actor));
+		return false;
+	}
+
+	const FName& LayerKey = GetLayerForActor(Actor);
+	if (!LayerNameToLBStrategy.Contains(LayerKey))
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy doesn't have a LBStrategy for Actor %s which is in Layer %s."), *AActor::GetDebugName(&Actor), *LayerKey.ToString());
+		return false;
+	}
+
+	// If this worker is not responsible for the Actor's layer, just return false.
+	if (VirtualWorkerIdToLayerName.Contains(LocalVirtualWorkerId) && VirtualWorkerIdToLayerName[LocalVirtualWorkerId] != GetLayerForActor(Actor))
+	{
+		return false;
+	}
+
+	return LayerNameToLBStrategy[LayerKey]->ShouldHaveAuthority(Actor);
+}
+
+VirtualWorkerId ULayeredLBStrategy::WhoShouldHaveAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy not ready to decide on authority for Actor %s."), *AActor::GetDebugName(&Actor));
+		return SpatialConstants::INVALID_VIRTUAL_WORKER_ID;
+	}
+
+	const FName& LayerKey = GetLayerForActor(Actor);
+	if (!LayerNameToLBStrategy.Contains(LayerKey))
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy doesn't have a LBStrategy for Actor %s which is in Layer %s."), *AActor::GetDebugName(&Actor), *LayerKey.ToString());
+		return SpatialConstants::INVALID_VIRTUAL_WORKER_ID;
+	}
+
+	const VirtualWorkerId ReturnedWorkerId = LayerNameToLBStrategy[LayerKey]->WhoShouldHaveAuthority(Actor);
+
+	UE_LOG(LogLayeredLBStrategy, Log, TEXT("LayeredLBStrategy returning virtual worker id %d for Actor %s."), ReturnedWorkerId, *AActor::GetDebugName(&Actor));
+	return ReturnedWorkerId;
+}
+
+SpatialGDK::QueryConstraint ULayeredLBStrategy::GetWorkerInterestQueryConstraint() const
+{
+	check(IsReady());
+	if (!VirtualWorkerIdToLayerName.Contains(LocalVirtualWorkerId))
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy doesn't have a LBStrategy for worker %d."), LocalVirtualWorkerId);
+		SpatialGDK::QueryConstraint Constraint;
+		Constraint.ComponentConstraint = 0;
+		return Constraint;
+	}
+	else
+	{
+		return LayerNameToLBStrategy[VirtualWorkerIdToLayerName[LocalVirtualWorkerId]]->GetWorkerInterestQueryConstraint();
+	}
+}
+
+FVector ULayeredLBStrategy::GetWorkerEntityPosition() const
+{
+	check(IsReady());
+	if (!VirtualWorkerIdToLayerName.Contains(LocalVirtualWorkerId))
+	{
+		return FVector{ 0.f, 0.f, 0.f };
+	}
+	else
+	{
+		return LayerNameToLBStrategy[VirtualWorkerIdToLayerName[LocalVirtualWorkerId]]->GetWorkerEntityPosition();
```

Comment:
Possibly need to check LayerNameToLBStrategy contains the LayorName.

Reply:
Good idea, done.